File: Mitchell_Shaeffer/mitchell.cpp
/*
   There are a total of 3 entries in the algebraic variable array.
   There are a total of 2 entries in each of the rate and state variable arrays.
   There are a total of 10 entries in the constant variable array.
 */
/*
 * VOI is time in component environment (ms).
 * ALGEBRAIC[0] is J_stim in component J_stim (per_ms).
 * CONSTANTS[0] is IstimStart in component J_stim (ms).
 * CONSTANTS[1] is IstimEnd in component J_stim (ms).
 * CONSTANTS[2] is IstimAmplitude in component J_stim (per_ms).
 * CONSTANTS[3] is IstimPeriod in component J_stim (ms).
 * CONSTANTS[4] is IstimPulseDuration in component J_stim (ms).
 * STATES[0] is Vm in component membrane (dimensionless).
 * ALGEBRAIC[1] is J_in in component J_in (per_ms).
 * ALGEBRAIC[2] is J_out in component J_out (per_ms).
 * CONSTANTS[5] is tau_in in component J_in (ms).
 * STATES[1] is h in component J_in_h_gate (dimensionless).
 * CONSTANTS[6] is tau_open in component J_in_h_gate (ms).
 * CONSTANTS[7] is tau_close in component J_in_h_gate (ms).
 * CONSTANTS[8] is V_gate in component J_in_h_gate (dimensionless).
 * CONSTANTS[9] is tau_out in component J_out (ms).
 * RATES[0] is d/dt Vm in component membrane (dimensionless).
 * RATES[1] is d/dt h in component J_in_h_gate (dimensionless).
 */

#include <iostream>
#include <iomanip>
#include <fstream>
#include <cmath>

using namespace std;

const int NALGEBRAICS = 3;
const int NRATES = 2;
const int NCONSTANTS = 10;

void initConsts(double* CONSTANTS, double* RATES, double *STATES)
{
    CONSTANTS[0] = 0;
    CONSTANTS[1] = 50000;
    CONSTANTS[2] = 0.2;
    CONSTANTS[3] = 500;
    CONSTANTS[4] = 1;
    STATES[0] = 0.00000820413566106744;
    CONSTANTS[5] = 0.3;
    STATES[1] = 0.8789655121804799;
    CONSTANTS[6] = 120.0;
    CONSTANTS[7] = 150.0;
    CONSTANTS[8] = 0.13;
    CONSTANTS[9] = 6.0;
}
void computeRates(double VOI, double* CONSTANTS, double* RATES, double* STATES, double* ALGEBRAIC)
{
    RATES[0] = ALGEBRAIC[1]+ALGEBRAIC[2]+ALGEBRAIC[0];
    RATES[1] = (STATES[0]<CONSTANTS[8] ? (1.00000 - STATES[1])/CONSTANTS[6] : - STATES[1]/CONSTANTS[7]);
}
void computeVariables(double VOI, double* CONSTANTS, double* RATES, double* STATES, double* ALGEBRAIC)
{
    ALGEBRAIC[0] = (VOI>=CONSTANTS[0]&&VOI<=CONSTANTS[1]&&(VOI - CONSTANTS[0]) -  floor((VOI - CONSTANTS[0])/CONSTANTS[3])*CONSTANTS[3]<=CONSTANTS[4] ? CONSTANTS[2] : 0.00000);
    ALGEBRAIC[1] = ( STATES[1]*( pow(STATES[0], 2.00000)*(1.00000 - STATES[0])))/CONSTANTS[5];
    ALGEBRAIC[2] = - (STATES[0]/CONSTANTS[9]);
}

void printSolution (ofstream &out, const double voi, const double STATES[])
{
    out << fixed << setprecision(8) << voi << " ";
    for (int i = 0; i < NRATES; i++)
        out << STATES[i] << " ";
    out << endl;
}

int main ()
{
    ofstream out("solution.dat");
    double dt = 0.1;
    double tmax = 500.0;
    int N = nearbyint(tmax / dt);

    double *ALGEBRAIC = new double[NALGEBRAICS];
    double *RATES = new double[NRATES];
    double *STATES = new double[NRATES];
    double *CONSTANTS = new double[NCONSTANTS]; 

    initConsts(CONSTANTS,RATES,STATES);
    for (int i = 0; i < N; i++)
    {
        double voi = i*dt;
        computeVariables(voi,CONSTANTS,RATES,STATES,ALGEBRAIC);
        computeRates(voi,CONSTANTS,RATES,STATES,ALGEBRAIC);
        
        printSolution(out,voi,STATES);
        for (int j = 0; j < NRATES; j++)
            STATES[j] = STATES[j] + RATES[j]*dt;
    }

    delete [] ALGEBRAIC;
    delete [] RATES;
    delete [] STATES;
    delete [] CONSTANTS;

    return 0;
}

File: Mitchell_Shaeffer/mitchell.py
# Size of variable arrays:
sizeAlgebraic = 3
sizeStates = 2
sizeConstants = 10

from math import *
from numpy import *

def createLegends():
    legend_states = [""] * sizeStates
    legend_rates = [""] * sizeStates
    legend_algebraic = [""] * sizeAlgebraic
    legend_voi = ""
    legend_constants = [""] * sizeConstants
    legend_voi = "time in component environment (ms)"
    legend_algebraic[0] = "J_stim in component J_stim (per_ms)"
    legend_constants[0] = "IstimStart in component J_stim (ms)"
    legend_constants[1] = "IstimEnd in component J_stim (ms)"
    legend_constants[2] = "IstimAmplitude in component J_stim (per_ms)"
    legend_constants[3] = "IstimPeriod in component J_stim (ms)"
    legend_constants[4] = "IstimPulseDuration in component J_stim (ms)"
    legend_states[0] = "Vm in component membrane (dimensionless)"
    legend_algebraic[1] = "J_in in component J_in (per_ms)"
    legend_algebraic[2] = "J_out in component J_out (per_ms)"
    legend_constants[5] = "tau_in in component J_in (ms)"
    legend_states[1] = "h in component J_in_h_gate (dimensionless)"
    legend_constants[6] = "tau_open in component J_in_h_gate (ms)"
    legend_constants[7] = "tau_close in component J_in_h_gate (ms)"
    legend_constants[8] = "V_gate in component J_in_h_gate (dimensionless)"
    legend_constants[9] = "tau_out in component J_out (ms)"
    legend_rates[0] = "d/dt Vm in component membrane (dimensionless)"
    legend_rates[1] = "d/dt h in component J_in_h_gate (dimensionless)"
    return (legend_states, legend_algebraic, legend_voi, legend_constants)

def initConsts():
    constants = [0.0] * sizeConstants; states = [0.0] * sizeStates;
    constants[0] = 0
    constants[1] = 50000
    constants[2] = 0.2
    constants[3] = 500
    constants[4] = 1
    states[0] = 0.00000820413566106744
    constants[5] = 0.3
    states[1] = 0.8789655121804799
    constants[6] = 120.0
    constants[7] = 150.0
    constants[8] = 0.13
    constants[9] = 6.0
    return (states, constants)

def computeRates(voi, states, constants):
    rates = [0.0] * sizeStates; algebraic = [0.0] * sizeAlgebraic
    rates[1] = custom_piecewise([less(states[0] , constants[8]), (1.00000-states[1])/constants[6] , True, -states[1]/constants[7]])
    algebraic[0] = custom_piecewise([greater_equal(voi , constants[0]) & less_equal(voi , constants[1]) & less_equal((voi-constants[0])-floor((voi-constants[0])/constants[3])*constants[3] , constants[4]), constants[2] , True, 0.00000])
    algebraic[1] = (states[1]*((power(states[0], 2.00000))*(1.00000-states[0])))/constants[5]
    algebraic[2] = -(states[0]/constants[9])
    rates[0] = algebraic[1]+algebraic[2]+algebraic[0]
    return(rates)

def computeAlgebraic(constants, states, voi):
    algebraic = array([[0.0] * len(voi)] * sizeAlgebraic)
    states = array(states)
    voi = array(voi)
    algebraic[0] = custom_piecewise([greater_equal(voi , constants[0]) & less_equal(voi , constants[1]) & less_equal((voi-constants[0])-floor((voi-constants[0])/constants[3])*constants[3] , constants[4]), constants[2] , True, 0.00000])
    algebraic[1] = (states[1]*((power(states[0], 2.00000))*(1.00000-states[0])))/constants[5]
    algebraic[2] = -(states[0]/constants[9])
    return algebraic

def custom_piecewise(cases):
    """Compute result of a piecewise function"""
    return select(cases[0::2],cases[1::2])

def solve_model():
    """Solve model with ODE solver"""
    from scipy.integrate import ode
    # Initialise constants and state variables
    (init_states, constants) = initConsts()

    # Set timespan to solve over
    voi = linspace(0,500, 500)

    # Construct ODE object to solve
    r = ode(computeRates)
    r.set_integrator('vode', method='bdf', atol=1e-06, rtol=1e-06, max_step=1)
    r.set_initial_value(init_states, voi[0])
    r.set_f_params(constants)

    # Solve model
    states = array([[0.0] * len(voi)] * sizeStates)
    states[:,0] = init_states
    for (i,t) in enumerate(voi[1:]):
        if r.successful():
            r.integrate(t)
            states[:,i+1] = r.y
        else:
            break

    # Compute algebraic variables
    algebraic = computeAlgebraic(constants, states, voi)
    return (voi, states, algebraic)

def plot_model(voi, states, algebraic):
    """Plot variables against variable of integration"""
    import pylab
    (legend_states, legend_algebraic, legend_voi, legend_constants) = createLegends()
    pylab.figure(1)
    #pylab.plot(voi,vstack((states,algebraic)).T)
    pylab.plot(voi,states[0])
    pylab.xlabel(legend_voi)
    #pylab.legend(legend_states + legend_algebraic, loc='best')
    pylab.show()

if __name__ == "__main__":
    (voi, states, algebraic) = solve_model()
    plot_model(voi, states, algebraic)